src/freenas/usr/local/lib/middlewared_truenas/plugins/failover.py

```diff
@@ -802,75 +700,144 @@ def _disabled_reasons(self, app):
-            except Exception as e:
-                self.logger.warn('Failed to set encryption key on standby node: %s', e)
-        return rv
+    @accepts(
+        Dict(
+            'update_encryption_keys',
+            Bool('sync_keys', default=True),
+            List(
+                'pools', items=[
+                    Dict(
+                        'pool_geli_keys',
+                        Str('name', required=True),
+                        Str('passphrase', required=True),
+                    )
+                ], default=[]
+            ),
+            List(
+                'datasets', items=[
+                    Dict(
+                        'dataset_keys',
+                        Str('name', required=True),
+                        Str('passphrase', required=True),
+                    )
+                ], default=[]
+            ),
+        )
+    )
+    async def update_encryption_keys(self, options):
+        if not options['pools'] and not options['datasets']:
+            raise CallError('Please specify pools/datasets to update')
+
+        async with ENCRYPTION_CACHE_LOCK:
+            keys = await self.encryption_keys()
+            for pool in options['pools']:
+                keys['geli'][pool['name']] = pool['passphrase']
+            for dataset in options['datasets']:
+                keys['zfs'][dataset['name']] = dataset['passphrase']
+            await self.middleware.call('cache.put', 'failover_encryption_keys', keys)
+            if options['sync_keys']:
+                await self.sync_keys_with_remote_node(lock=False)
 
     @private
-    @accepts()
-    def encryption_clearkey(self):
-        # FIXME: we could get rid of escrow, middlewared can do that job
-        with LocalEscrowCtl() as escrowctl:
-            return escrowctl.clear()
+    @accepts(
+        Dict(
+            'remove_encryption_keys',
+            Bool('sync_keys', default=True),
+            List('pools', items=[Str('pool')], default=[]),
+            List('datasets', items=[Str('dataset')], default=[]),
+        )
+    )
+    async def remove_encryption_keys(self, options):
+        if not options['pools'] and not options['datasets']:
+            raise CallError('Please specify pools/datasets to remove')
+
+        async with ENCRYPTION_CACHE_LOCK:
+            keys = await self.encryption_keys()
+            for pool in options['pools']:
+                keys['geli'].pop(pool, None)
+            for dataset in options['datasets']:
+                keys['zfs'] = {
+                    k: v for k, v in keys['zfs'].items() if k != dataset and not k.startswith(f'{dataset}/')
+                }
+            await self.middleware.call('cache.put', 'failover_encryption_keys', keys)
+            if options['sync_keys']:
+                await self.sync_keys_with_remote_node(lock=False)
 
     @private
     @job()
     def encryption_attachall(self, job):
```

Comment:
Since you're here and making all these wonderful changes, it would be best to name this function to something not so ambiguous. :)

Maybe something like `def attach_all_geli_providers`